lib/AST/ASTPrinter.cpp

```diff
@@ -1901,16 +1902,27 @@ void PrintAST::printBodyIfNecessary(const AbstractFunctionDecl *decl) {
   printBraceStmt(decl->getBody(), /*newlineIfEmpty*/!isa<AccessorDecl>(decl));
 }
 
-void PrintAST::printMutabilityModifiersIfNeeded(const FuncDecl *FD) {
+void PrintAST::printSelfAccessKindModifiersIfNeeded(const FuncDecl *FD) {
+  if (!Options.PrintSelfAccessKindKeyword)
+    return;
+
   const auto *AD = dyn_cast<AccessorDecl>(FD);
 
-  if (FD->isMutating()) {
-    if (AD == nullptr || AD->isAssumedNonMutating())
-      if (!Options.excludeAttrKind(DAK_Mutating))
-        Printer.printKeyword("mutating", Options, " ");
-  } else if (AD && AD->isExplicitNonMutating() &&
-             !Options.excludeAttrKind(DAK_Mutating)) {
-    Printer.printKeyword("nonmutating", Options, " ");
+  switch (FD->getSelfAccessKind()) {
+  case SelfAccessKind::Mutating:
+    if ((!AD || AD->isAssumedNonMutating()) &&
+        !Options.excludeAttrKind(DAK_Mutating))
+      Printer.printKeyword("mutating", Options, " ");
+    break;
+  case SelfAccessKind::NonMutating:
+    if (AD && AD->isExplicitNonMutating() &&
+        !Options.excludeAttrKind(DAK_NonMutating))
```

Comment:
This was checking `DAK_Mutating` before, which seems like a latent bug, but I don't think it's observable since the only place I found that adds `DAK_Mutating` but not `DAK_NonMutating` is protocol conformance diagnostics for non-accessors, but this code only prints `nonmutating` on accessors.  Oh well, at least it's correct now :shrug: